pkg/apis/metrics/types.go

```diff
@@ -16,14 +16,205 @@ limitations under the License.
+	// List of raw pod metrics.
+	Items []RawPodMetrics `json:"items"`
+}
+
+// RawContainerMetrics holds container-level unprocessed sample metrics.
+type RawContainerMetrics struct {
+	// Reference to the measured container.
+	Name string `json:"name"`
+	// Metadata labels associated with this container (not Kubernetes labels).
+	// For example, docker labels.
+	Labels map[string]string `json:"labels,omitempty"`
+	// Historical metric samples gathered from the container.
+	Samples []ContainerSample `json:"samples" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+	// Application-specific metrics.
+	CustomMetrics []CustomMetric `json:"customMetrics,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// NonLocalObjectReference contains enough information to locate the referenced object.
+type NonLocalObjectReference struct {
+	Name      string    `json:"name"`
+	Namespace string    `json:"namespace"`
+	UID       types.UID `json:"uid,omitempty"`
+}
+
+// Sample defines metadata common to all sample types.
+// Samples may not be nested within other samples.
+type Sample struct {
+	// The time this data point was collected at.
+	SampleTime unversioned.Time `json:"sampleTime"`
+}
+
+// AggregateSample contains a metric sample point of data aggregated across containers.
+type AggregateSample struct {
+	Sample `json:",inline"`
+	// Metrics pertaining to CPU resources.
+	CPU *CPUMetrics `json:"cpu,omitempty"`
+	// Metrics pertaining to memory (RAM) resources.
+	Memory *MemoryMetrics `json:"memory,omitempty"`
+	// Metrics pertaining to network resources.
+	Network *NetworkMetrics `json:"network,omitempty"`
+}
+
+// PodSample contains a metric sample point of pod-level resources.
+type PodSample struct {
+	Sample `json:",inline"`
+	// Metrics pertaining to network resources.
+	Network *NetworkMetrics `json:"network,omitempty"`
+}
+
+// ContainerSample contains a metric sample point of container-level resources.
+type ContainerSample struct {
+	Sample `json:",inline"`
+	// Metrics pertaining to CPU resources.
+	CPU *CPUMetrics `json:"cpu,omitempty"`
+	// Metrics pertaining to memory (RAM) resources.
+	Memory *MemoryMetrics `json:"memory,omitempty"`
+}
+
+// NetworkMetrics contains data about network resources.
+type NetworkMetrics struct {
+	// Cumulative count of bytes received.
+	RxBytes *resource.Quantity `json:"rxBytes,omitempty"`
+	// Cumulative count of receive errors encountered.
+	RxErrors *int64 `json:"rxErrors,omitempty"`
+	// Cumulative count of bytes transmitted.
+	TxBytes *resource.Quantity `json:"txBytes,omitempty"`
+	// Cumulative count of transmit errors encountered.
+	TxErrors *int64 `json:"txErrors,omitempty"`
+}
+
+// CPUMetrics contains data about CPU usage.
+type CPUMetrics struct {
+	// Total CPU usage (sum of all cores) averaged over the sample window.
+	// The "core" unit can be interpreted as CPU core-seconds per second.
+	TotalCores *resource.Quantity `json:"totalCores,omitempty"`
```

Comment:
I'm trying to avoid duplicating data that's provided by the kubernetes API. In this case, number of cores is provided by Capacity (will be [Allocatable](https://github.com/kubernetes/kubernetes/pull/17201)).